compiler/rustc_trait_selection/src/traits/coherence.rs

```diff
@@ -675,19 +700,148 @@ pub enum OrphanCheckErr<'tcx> {
-    debug!(?trait_ref);
+    let trait_ref = tcx.impl_trait_ref(impl_def_id).unwrap();
+    debug!(trait_ref = ?trait_ref.skip_binder());
 
     // If the *trait* is local to the crate, ok.
-    if trait_ref.def_id.is_local() {
-        debug!("trait {:?} is local to current crate", trait_ref.def_id);
-        return Ok(());
+    if let Some(def_id) = trait_ref.skip_binder().def_id.as_local() {
+        debug!("trait {def_id:?} is local to current crate");
+        return OrphanCheckResult::Ok;
+    }
+
+    let infcx = tcx.infer_ctxt().intercrate(true).build();
+    let cause = ObligationCause::dummy();
+    let args = infcx.fresh_args_for_item(cause.span, impl_def_id);
+    let trait_ref = trait_ref.instantiate(tcx, args);
+
+    match orphan_check_trait_ref::<!>(trait_ref, InCrate::Local, |user_ty| {
+        let ty::Alias(..) = user_ty.kind() else { return Ok(user_ty) };
+
+        let ocx = ObligationCtxt::new(&infcx);
+        let ty = ocx.normalize(&cause, ty::ParamEnv::empty(), user_ty);
+        let ty = infcx.resolve_vars_if_possible(ty);
+        let errors = ocx.select_where_possible();
+        if !errors.is_empty() {
+            return Ok(user_ty);
+        }
+
+        let ty = if infcx.next_trait_solver() {
+            let mut fulfill_cx = <dyn TraitEngine<'_>>::new(&infcx);
+            infcx
+                .at(&cause, ty::ParamEnv::empty())
+                .structurally_normalize(ty, &mut *fulfill_cx)
+                .map(|ty| infcx.resolve_vars_if_possible(ty))
+                .unwrap_or(ty)
+        } else {
+            ty
+        };
+
+        Ok(ty)
+    }) {
+        Ok(OrphanCheckResult::Ok) => OrphanCheckResult::Ok,
+        Ok(OrphanCheckResult::Issue99554(UncoveredTyParams { uncovered, local_ty })) => {
+            let mut collector =
+                UncoveredTyParamCollector { infcx: &infcx, uncovered_params: Default::default() };
+            uncovered.visit_with(&mut collector);
+            debug_assert!(!collector.uncovered_params.is_empty()); // FIXME: possibly reachable
+
+            OrphanCheckResult::Issue99554(UncoveredTyParams {
+                uncovered: collector.uncovered_params,
+                local_ty,
+            })
+        }
+        Ok(OrphanCheckResult::Err(OrphanCheckErr::UncoveredTyParams(UncoveredTyParams {
+            uncovered,
+            local_ty,
+        }))) => {
+            let mut collector =
+                UncoveredTyParamCollector { infcx: &infcx, uncovered_params: Default::default() };
+            uncovered.visit_with(&mut collector);
+            debug_assert!(!collector.uncovered_params.is_empty()); // FIXME: possibly reachable
+
+            OrphanCheckResult::Err(OrphanCheckErr::UncoveredTyParams(UncoveredTyParams {
+                uncovered: collector.uncovered_params,
+                local_ty,
+            }))
+        }
+        Ok(OrphanCheckResult::Err(OrphanCheckErr::NonLocalInputType(tys))) => {
```

Comment:
not using `FmtPrinter` since remapping of ty vars now happens here, not over at the diagnostic code. Furthermore, the diagnostic code in hir analysis does further `Ty` processing so we can't fmt it here to a `String`.

Note that doing the remapping here allows us to not pass around an `InferCtxt` / `Box<_>` from trait selection to hir analysis which was super gross. You can see the perks of the new structuring by cmp'ing commit 2 and 3. I'm so happy I've managed to get rid of that.